### PR TITLE
feat: user can skip upstream when select service_id

### LINKED ISSUE #1213: The process of creating routes using dashboard needs improvement
# Bug report

Now, upstream is necessary when using dashboard to create a route. If service is selected, the user needs to fill in upstream twice, which is unreasonable. I drew a flowchart to express what I mean.

image

### LINKED ISSUE #1278: bug(fe): route created with service can not use the service upstream
# Bug report

## Describe the bug

Refer to the [description](https://github.com/apache/apisix/blob/master/doc/architecture-design.md#service), when create a route with service id, this route will have the same upstream and plugin configuration.

When I try to create a route with service in dashboard, although I have selected a service in the first step, I still  have to define a new`upstream` because of the required validation in the second step, and  at the same time, the new `upstream` definition will cover the service's upstream.

## How to Reproduce

1. create a service
2. create a route, select the service in the first step, click `Next` button
3. in the second step, we have to define the upstream

## Expected behavior

I would be better to use service data to auto fill in the `upstream` form in the second step.

## System information

- OS: Fedora 32
- Browser (if applies) Chrome
- Version: master branch

### PR DESCRIPTION
**Pull request by LiteSun**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bugfix
- [ ] New feature provided
- [x] Improve performance
- [ ] Backport patches

- Related issues
resolve #1213, resolve #1278 
___
### New feature or improvement
- Describe the details and related test reports.
user can skip upstream when select service_id.
___
### Please add the corresponding test cases if necessary.
- [ ] Todo
___

### CODE REVIEW

**Comment by liuxiran** [comment]
just to make sure that this pr could also solves #1278 , if so I will move to other issues, thanks ^-^

cc @LiteSun @juzhiyuan

**Comment by LiteSun** [comment]
> just to make sure that this pr could also solves #1278 , if so I will move to other issues, thanks ^-^
> 
> cc @LiteSun @juzhiyuan

According to the issue(#1278) description, this PR will resolve the issue. 😄

File: web/src/pages/Route/components/CreateStep4/CreateStep4.tsx [hunk 1]
```diff
@@ -50,7 +50,7 @@ const CreateStep4: React.FC<Props> = ({ form1, form2, redirect, upstreamRef, ...
           <h2 style={style}>
             {formatMessage({ id: 'page.route.steps.stepTitle.defineApiBackendServe' })}
           </h2>
-          <Step2 form={form2} upstreamRef={upstreamRef} disabled />
+          <Step2 form={form2} upstreamRef={upstreamRef} disabled isServiceId={form1.getFieldValue('service_id') !== ''} />
```
[the remark this answers is not in the document]
Comment (juzhiyuan) [reply web/src/pages/Route/components/CreateStep4/CreateStep4.tsx:53]: or `serviceExist`

[the remark this answers is not in the document]
Comment (LiteSun) [reply web/src/pages/Route/components/CreateStep4/CreateStep4.tsx:53]: ok

File: web/cypress/integration/route/create-route-can-skip-upstream.spec.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    routeName: 'test_route',
+    ip: '127.0.0.1',
+  }
+  const domSelector = {
+    name: '#name',
+    nodes_0_host: '#nodes_0_host',
+    upstream_id: '#upstream_id',
+    input: ':input',
+    custom: '[title=Custom]',
+    titleName: '[title=Name]',
```
Comment (juzhiyuan) [inline web/cypress/integration/route/create-route-can-skip-upstream.spec.js:32]: we may use sanke_case in the future.

File: web/cypress/integration/route/create-route-can-skip-upstream.spec.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    cy.get(domSelector.name).type(data.upstreamName);
+    cy.get(domSelector.nodes_0_host).type(data.ip);
+    cy.contains('Next').click();
+    cy.contains('Submit').click();
+
+    cy.visit('/');
+    cy.contains('Service').click();
+    cy.contains('Create').click();
+    cy.get(domSelector.name).type(data.serviceName);
+    cy.get(domSelector.custom).click();
```
Comment (juzhiyuan) [inline web/cypress/integration/route/create-route-can-skip-upstream.spec.js:54]: custom what? custom upstream here.

File: web/cypress/integration/route/create-route-can-skip-upstream.spec.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    cy.contains(data.upstreamName).click();
+    cy.contains('Next').click();
+    cy.contains('Next').click();
+    cy.contains('Submit').click();
```
Comment (juzhiyuan) [inline web/cypress/integration/route/create-route-can-skip-upstream.spec.js:58]: Need to check if those two resources are created.

File: web/cypress/integration/route/create-route-can-skip-upstream.spec.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  });
+
+  it('should select service_id can skip upstream when create route', () => {
```
Comment (juzhiyuan) [inline web/cypress/integration/route/create-route-can-skip-upstream.spec.js:61]:
```suggestion
  it('should skip upstream module after service is selected when creating route', () => {
```

File: web/cypress/integration/route/create-route-can-skip-upstream.spec.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    cy.visit('/');
+    cy.contains('Route').click();
+    cy.contains('Create').click();
+
+    // When no select service_id, the option None does not exist
```
Comment (juzhiyuan) [inline web/cypress/integration/route/create-route-can-skip-upstream.spec.js:66]:
```suggestion
    // The None option doesn't exist when service isn't selected
```

File: web/cypress/integration/route/create-route-can-skip-upstream.spec.js [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    cy.get(domSelector.custom).click();
+    cy.contains('None').should('not.exist');
+
+    cy.contains('Previous').click();
+    cy.contains('None').click();
+    cy.contains(data.serviceName).click();
+    cy.contains('Next').click();
+
+    // make sure upstream data can be saved
+    cy.get(domSelector.custom).click();
```
Comment (juzhiyuan) [inline web/cypress/integration/route/create-route-can-skip-upstream.spec.js:78]: L72-L75 is creating a route with Service, why we still need to set upstream?

File: web/cypress/integration/route/create-route-can-skip-upstream.spec.js [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+    cy.contains(data.upstreamName).click();
+    cy.contains('None').click();
+    cy.contains('Next').click();
+    cy.contains('Next').click();
+    cy.contains('Submit').click();
+    cy.contains('Goto List').click();
+  });
+
+  it('should select service_id can skip upstream when edit route', () => {
```
Comment (juzhiyuan) [inline web/cypress/integration/route/create-route-can-skip-upstream.spec.js:90]:
```suggestion
  it('should skip Upstream module after service is selected when editing route', () => {
```

File: web/cypress/integration/route/create-route-can-skip-upstream.spec.js [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    cy.contains(data.routeName).siblings().contains('Edit').click();
+    cy.get(domSelector.testService).click();
+    cy.contains('None').click();
+    cy.contains('Next').click();
+    cy.get(domSelector.upstream_id).click();
+    cy.contains('None').should('not.exist');
+    cy.contains(data.upstreamName).click();
+    cy.contains('Next').click();
+    cy.contains('Next').click();
+    cy.contains('Submit').click();
```
Comment (juzhiyuan) [inline web/cypress/integration/route/create-route-can-skip-upstream.spec.js:105]: Edit successfully?

File: web/cypress/integration/route/create-route-can-skip-upstream.spec.js [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+  });
+
+  it('should delete upstream service and route', () => {
+    cy.visit('/');
+    cy.contains('Upstream').click();
+    cy.contains(data.upstreamName).siblings().contains('Delete').click();
+    cy.contains('button', 'Confirm').click();
+
```
Comment (juzhiyuan) [inline web/cypress/integration/route/create-route-can-skip-upstream.spec.js:113]: successfully?

File: web/cypress/integration/route/create-route-can-skip-upstream.spec.js [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    cy.contains('Next').click();
+    cy.get(domSelector.upstream_id).click();
+    cy.contains('None').should('not.exist');
+    cy.contains(data.upstreamName).click();
+    cy.contains('Next').click();
+    cy.contains('Next').click();
+    cy.contains('Submit').click();
+  });
+
+  it('should delete upstream service and route', () => {
```
Comment (juzhiyuan) [inline web/cypress/integration/route/create-route-can-skip-upstream.spec.js:108]:
```suggestion
  it('should delete upstream, service and route', () => {
```

**Comment by LiteSun** [comment]
Ping

**Review by juzhiyuan — APPROVED** [review]

**Review by gxthrj — APPROVED** [review]

**Pull request merged**